### PR TITLE
stdlib: expand annotations, pickle, weakref, and functools compatibility

### PR DESCRIPTION
**Pull request by youknowone**
## Summary

- complete lazy annotation introspection and preserve class annotation namespaces
- support dict subclasses and slotted dict subclasses in interpreter paths used by stdlib
- align pickle global import failures with CPython behavior
- finish typing and weakref edge cases, including dataclass slots and GC-sensitive caches
- preserve keyword calls and accelerator-compatible behavior through functools

## Validation

- cargo check --features dynasm
- cargo test --features dynasm
- typing, weakref, and functools snippets with interpreter, default JIT, and dynasm JIT
- test_dataclasses: 276 passed, 2 skipped
- TestPartialPy: 34 passed
- TestCmpToKeyC: 8 passed, 1 skipped
- TestReduceC: 3 passed
- dynasm benchmark suite: 11/11 passed

### CODE REVIEW

File: pyre/pyre-interpreter/src/eval.rs [hunk 1]
```diff
@@ -3354,27 +3375,21 @@ impl OpcodeStepExecutor for PyFrame {
...
-        let w_globals = self.get_w_globals();
-        if !w_globals.is_null() {
-            if let Some(val) =
-                unsafe { pyre_object::dictmultiobject::w_dict_getitem_str(w_globals, name) }
-            {
-                self.push(val);
+    fn load_from_dict_or_globals(&mut self, name: &str, nameindex: usize) -> Result<(), PyError> {
+        let mapping = self.pop();
+        let key = pyre_object::w_str_new(name);
+        match crate::baseobjspace::getitem(mapping, key) {
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/eval.rs:3392]: Fixed in follow-up #718. The residual now uses mapping subscription, falls back only on KeyError, propagates other mapping exceptions, and resolves globals through the same mapping-aware path. A forced-JIT annotation thunk regression covers this.

File: pyre/pyre-interpreter/src/baseobjspace.rs [hunk 2]
```diff
@@ -5070,6 +5070,100 @@ unsafe fn type_getattr_hook_or_err(
...
+            unsafe { w_type_get_name(obj) },
+        )));
+    }
+    if !unsafe { is_none(value) } && !callable_w(value) {
+        return Err(PyError::type_error(
+            "__annotate__ must be callable or None",
+        ));
+    }
+    crate::type_dict_store(obj, "__annotate_func__", value);
+    crate::type_dict_delete(obj, "__annotations_cache__");
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/baseobjspace.rs:5443]: Fixed in follow-up #718. Assigning `None` to `__annotate__` now preserves an already materialized `__annotations_cache__`; a new callable still invalidates it.

File: pyre/pyre-interpreter/src/baseobjspace.rs [hunk 3]
```diff
@@ -5070,6 +5070,100 @@ unsafe fn type_getattr_hook_or_err(
...
+}
+
+pub(crate) fn type_set_annotations(obj: PyObjectRef, value: PyObjectRef) -> PyResult {
+    if !unsafe { pyre_object::w_type_is_heaptype(obj) } {
+        return Err(PyError::type_error(format!(
+            "cannot set '__annotations__' attribute of immutable type '{}'",
+            unsafe { w_type_get_name(obj) },
+        )));
+    }
+    crate::type_dict_store(obj, "__annotations_cache__", value);
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/baseobjspace.rs:5421]: Fixed in follow-up #718. The setter now updates an explicit class-body `__annotations__` entry in place (matching CPython 3.14), otherwise it writes `__annotations_cache__`; deletion independently clears either representation.

File: pyre/pyre-interpreter/src/baseobjspace.rs [hunk 4]
```diff
@@ -5373,6 +5373,98 @@ unsafe fn type_getattr_hook_or_err(
...
+    if !unsafe { pyre_object::w_type_is_heaptype(obj) } {
+        return Err(PyError::type_error(format!(
+            "cannot set '__annotate__' attribute of immutable type '{}'",
+            unsafe { w_type_get_name(obj) },
+        )));
+    }
+    if !unsafe { is_none(value) } && !callable_w(value) {
+        return Err(PyError::type_error("__annotate__ must be callable or None"));
+    }
+    crate::type_dict_store(obj, "__annotate_func__", value);
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/baseobjspace.rs:5442]: Addressed in follow-up #718 with CPython 3.14 parity. CPython intentionally leaves an explicit class-body `__annotate__` key visible when the descriptor's internal `__annotate_func__` is assigned; it does not overwrite that public key. The getter now gives the explicit key precedence, and regression coverage pins this behavior.

File: pyre/pyre-interpreter/src/pyframe.rs [hunk 5]
```diff
@@ -3697,24 +3697,10 @@ pub fn createframe_obj(
...
-        if backing.is_null() {
-            w_globals
-        } else {
-            backing
-        }
-    };
+    // `pyframe.py:49 self.w_globals = w_globals`: preserve the exact object.
+    // This is observable for dict subclasses (notably annotationlib's
+    // `_StringifierDict`, whose `__missing__` creates ForwardRef values) and
+    // is the identity MAKE_FUNCTION must pass on to each callee frame.
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/pyframe.rs:3703]: Addressed in follow-up #718, with one parity correction to the suggested implementation. CPython 3.14 bypasses a dict subclass's Python-level `__delitem__` for `DELETE_GLOBAL`, so routing through generic `baseobjspace::delitem` would be observably wrong. Pyre now resolves the composed dict-subclass backing and deletes through its intrinsic dict strategy, which fixes the layout issue while preserving CPython/PyPy semantics.

File: pyre/pyre-jit/src/jit/codewriter.rs [hunk 6]
```diff
@@ -6640,10 +6640,22 @@ impl CodeWriter {
...
+                // the synthetic null return.  Leaving this block with no exit
+                // would instead make `flatten.py:107-109` mistake its full
+                // FrameState input tuple for return arguments.
+                let abort_return = super::flow::Link::new(
+                    vec![super::flow::Constant::none().into()],
+                    Some(graph.returnblock.clone()),
+                    None,
+                )
+                .into_ref();
+                append_exit(&current_block.block(), abort_return);
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-jit/src/jit/codewriter.rs:6658]: Fixed in follow-up #718. The large-CALL bailout now immediately continues the walker after `emit_abort_permanent!`, so no synthetic result or later operation is recorded on the closed block. A two-consecutive-15-argument-call regression is included.

File: pyre/pyre-interpreter/src/typedef.rs [hunk 7]
```diff
@@ -1935,7 +1935,11 @@ fn module_annotations_get(args: &[PyObjectRef]) -> Result<PyObjectRef, crate::Py
                     "__annotate__ returned non-dict of type '{received}'"
                 )));
             }
-            result
+            // Module annotation functions consult the mutable
+            // `__conditional_annotations__` set.  A partially executed
+            // module may therefore produce a larger dict on a later read;
+            // do not freeze the intermediate result in `__annotations__`.
+            return Ok(result);
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/typedef.rs:1942]: Fixed in follow-up #718. A successful module annotation thunk result now reaches the common `__annotations__` store path and is cached in the module dictionary.

File: pyre/pyre-interpreter/src/call.rs [hunk 8]
```diff
@@ -3826,6 +3842,16 @@ fn build_class_inner(
         }
     }
 
+    if let Some(classdictcell_root) = classdictcell_root {
+        if unsafe { pyre_object::is_type(w_type) } {
+            let classdictcell = pyre_object::gc_roots::shadow_stack_get(classdictcell_root);
+            let type_dict = unsafe { pyre_object::w_type_get_dict_ptr(w_type) as PyObjectRef };
+            if !type_dict.is_null() {
+                unsafe { pyre_object::w_cell_set(classdictcell, type_dict) };
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/call.rs:3850]: Fixed in follow-up #718 for both the default `__build_class__` path and `type.__new__`: `__classdict__` is bound to the completed type dictionary before any `__set_name__` callback. Regressions cover both ordinary and custom-metaclass classes.

File: pyre/pyre-interpreter/src/module/_pickle/pickler.rs [hunk 9]
```diff
@@ -1891,7 +1891,29 @@ fn whichmodule(w_obj: PyObjectRef, name: &str) -> Result<PyObjectRef, PyError> {
...
+    let resolved = match crate::module::_pickle::try_resolve_global(&module_name, name, true) {
+        Ok(value) => value,
+        Err(error)
+            if matches!(
+                error.kind,
+                crate::PyErrorKind::ImportError
+                    | crate::PyErrorKind::ModuleNotFoundError
+                    | crate::PyErrorKind::ValueError
+                    | crate::PyErrorKind::KeyError
+            ) =>
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/module/_pickle/pickler.rs:1907]: Fixed in follow-up #718. Import/sys.modules failures are translated separately; dotted attribute traversal happens afterward, so a real module's PEP 562 `__getattr__` KeyError propagates unchanged.
